go/pkg/chunker/chunker.go

```diff
@@ -139,65 +137,47 @@ func (c *Chunker) Next() (*Chunk, error) {
 		return nil, ErrEOF
 	}
 	c.initialized = true
-	bytesLeft := c.bytesLeft()
 	if c.digest.Size == 0 {
 		return emptyChunk, nil
 	}
-	toRead := IOBufferSize
-	if int64(toRead) > bytesLeft {
-		toRead = int(bytesLeft)
-	}
-	if c.contents != nil {
-		c.buffer = c.contents
-	} else if c.reader == nil {
+	if c.contents == nil && c.reader == nil {
 		f, err := os.Open(c.path)
 		if err != nil {
 			return nil, err
 		}
-		c.reader = f
-		if c.buffer == nil {
-			c.buffer = make([]byte, toRead)
-			if c.offset == 0 && int64(toRead) == c.digest.Size {
-				c.contents = c.buffer
-			}
-		}
+		c.reader = bufio.NewReaderSize(f, IOBufferSize)
 	}
-	start := int(c.offset % int64(len(c.buffer)))
+	bytesLeft := c.bytesLeft()
 	bytesToSend := c.chunkSize
 	if bytesLeft < int64(bytesToSend) {
 		bytesToSend = int(bytesLeft)
 	}
 	var data []byte
-	leftover := bytesToSend
 	if c.reader != nil {
```

Comment:
I would structure this as follows to avoid the nested branching logic:
buffOffset := c.offset
if c.reader != nil {
  buffOffset = 0
  newBuffSize := bytesToSend
  if c.offset == 0 && c.digest.Size <= int64(IOBufferSize) {
   newBuffSize = c.digest.Size
  }
  c.contents = make([]byte, newBuffSize)
  n, err := io.ReadFull(c.reader, c.contents)
  ... // error handling
}
data = c.contents[buffOffset:buffOffset + bytesToSend]
...

Reply:
That doesn't quite work -- I can't set c.contents unless it's the fully cached data. And you set it unconditionally here whenever reader is nil. So the next read would fail here.